lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/AvroPageDataReader.java

```diff
@@ -0,0 +1,1046 @@
+    private Schema writerSchema;
+    private final PageBuilder pageBuilder;
+    private RowBlockBuildingDecoder rowBlockBuildingDecoder;
+    private final AvroTypeManager typeManager;
+
+    public AvroPageDataReader(Schema readerSchema, AvroTypeManager typeManager)
+            throws AvroTypeException
+    {
+        this.readerSchema = requireNonNull(readerSchema, "readerSchema is null");
+        writerSchema = this.readerSchema;
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        try {
+            Type readerSchemaType = typeFromAvro(this.readerSchema, typeManager);
+            verify(readerSchemaType instanceof RowType, "Root Avro type must be a row");
+            pageBuilder = new PageBuilder(readerSchemaType.getTypeParameters());
+            initialize();
+        }
+        catch (org.apache.avro.AvroTypeException e) {
+            throw new AvroTypeException(e);
+        }
+    }
+
+    private void initialize()
+            throws AvroTypeException
+    {
+        verify(readerSchema.getType() == Schema.Type.RECORD, "Avro schema for page reader must be record");
+        verify(writerSchema.getType() == Schema.Type.RECORD, "File Avro schema for page reader must be record");
+        rowBlockBuildingDecoder = new RowBlockBuildingDecoder(writerSchema, readerSchema, typeManager);
+    }
+
+    @Override
+    public void setSchema(Schema schema)
+    {
+        requireNonNull(schema, "schema is null");
+        if (schema != writerSchema) {
+            writerSchema = schema;
+            try {
+                initialize();
+            }
+            catch (org.apache.avro.AvroTypeException e) {
+                throw new UncheckedAvroTypeException(new AvroTypeException(e));
+            }
+            catch (AvroTypeException e) {
+                throw new UncheckedAvroTypeException(e);
+            }
+        }
+    }
+
+    @Override
+    public Optional<Page> read(Optional<Page> ignoredReuse, Decoder decoder)
+            throws IOException
+    {
+        Optional<Page> page = Optional.empty();
+        rowBlockBuildingDecoder.decodeIntoPageBuilder(decoder, pageBuilder);
+        if (pageBuilder.isFull()) {
+            page = Optional.of(pageBuilder.build());
+            pageBuilder.reset();
+        }
+        return page;
+    }
+
+    public Optional<Page> flush()
+    {
+        if (!pageBuilder.isEmpty()) {
+            Optional<Page> lastPage = Optional.of(pageBuilder.build());
+            pageBuilder.reset();
+            return lastPage;
+        }
+        return Optional.empty();
+    }
+
+    private abstract static class BlockBuildingDecoder
+    {
+        protected abstract void decodeIntoBlock(Decoder decoder, BlockBuilder builder)
+                throws IOException;
+    }
+
+    private static BlockBuildingDecoder createBlockBuildingDecoderForAction(Resolver.Action action, AvroTypeManager typeManager)
+            throws AvroTypeException
```

Comment:
I probably wouldn't recommend using the resolver. The only thing that needs to be resolved is the order to read struct fields, skip, and then how to fill in defaults. That's actually pretty simple and can be done directly. Check out our [implementation in Python](https://github.com/apache/iceberg/blob/master/python/pyiceberg/avro/resolver.py#L141-L156).

Basically, that creates a list of (optional position, reader) pairs by doing the following:

1. Loop over the file schema's record and create a pair for every field
    - If there is a corresponding field in the read schema, use that field's position from the read schema
    - If the field is not in the read schema, use empty
2. Loop over the read schema and add any needed pairs
    - If the field was in the file schema, skip it
    - If the field was missing, create a pair from the field's position in the read schema and a constant reader with the field's default

Then to read, you just loop over those pairs. If the position is present, get the value from the reader and set that position in the record. If the position is not present, call skip to consume and discard the bytes.

I think this is overall way cleaner, although I like what you're doing here and using the resolver to create your reader tree, rather than resolving in each record.

Reply:
Oh that's a good point. I definitely don't love the data model of the Actions returned for this or other use cases. When I was implementing I was just kinda peeled back layers of Generic Datum Reader and FastDatumReader until I didn't need to anymore and just pretty much copied the use from those classes. The reason I might keep using it is that is also provides Reader and Writer Union resolution steps that I originally wan't super confident on implementing but need (I imagine I could but would need to think on it, I think it would involve load bearing exceptions). From my understanding after looking at the pyiceberg code, Iceberg doesn't support non-trivial(optional) unions and doesn't need to do union resolution correct?

Reply:
Do you want to address this now or in a follow up?

Reply:
I'll ping @rdblue offline and we'll get this sorted.